ENH: distutils: make msvc + mingw-gfortran work

xoviat opened the pull request:
Enable numpy.distutils to link mingw-w64-gfortran compiled files to MSVC objects, 
by putting everything gfortran to separate DLL files to avoid CRT issues.

On installation, the extra DLL files are stuffed to an `extra-dll` subdirectory
of the package, which is added to PATH so that Windows can load the files.
The filenames of the libraries contain a sha1 hash of the input object files,
so DLL hell should be avoided (barring sha1 hash collisions).

This is a bit of a hack, but it is limited to the "msvc/gnu95" special case.
Moreover, in addition to mingwpy (which may still need further work?),
this is the currently the only working easily available compiler combination 
to build Python.org compatible binaries for Windows.

To handle static libraries, we need to add a concept of a "fake" static library
to numpy.distutils. This is just a collection of object files whose filenames are
listed in `*.cobjects` and `*.fobjects` files. This is necessary, because MSVC
cannot build linkable static libraries out of gfortran objects, so the linking
cannot be done in the `build_clib` stage but has to be postponed to `build_ext`.
(And in distutils, the only good way to share information between stages
is to store it on disk).

Limitations: fortran code that needs symbols from C code provided by the
project cannot be compiled like --- this would require using <s>two C compilers
(gcc + msvc) and this is getting too complicated</s> import libraries in also
to the other direction, so it's not supported currently (will fail with undefined 
symbol errors  at link time). However, the typical case of wrapping Fortran 
codes to Python doesn't need this.

Moreover, if linking against BLAS/LAPACK, that must be provided in
a gfortran-compatible import library. The same machinery then will 
automatically build a wrapper DLL for it for MSVC if needed.

Scipy Appveyor builds & wheels enabled by this: 
https://github.com/scipy/scipy/pull/7616
https://ci.appveyor.com/project/scipy/scipy

matthew-brett commented:
First of a few cross-references to comments in the previous PR:

"""
@rgommers The other issue that I am unsure how to address is how to gracefully handle DLL dependencies . Do you have any ideas on this?"
""" 

@xoviat - I guess you mean, for each scipy extension module, telling disutils which Fortran DLLs they depend on?

matthew-brett commented:
Comments on mechanism for specifying library order:

* https://github.com/numpy/numpy/pull/9429#issuecomment-315686123
* https://github.com/numpy/numpy/pull/9429#issuecomment-315686728
* https://github.com/numpy/numpy/pull/9429#issuecomment-315687097

Approach here is to list all the scipy Fortran libraries inside numpy distutils.  It would be better to specify them inside Scipy, perhaps (@pv) in setup.py.  But we need to take care of ordering of libraries.  Do I understand right, that this will need changes to numpy distutils?

xoviat commented:
>  @xoviat - I guess you mean, for each scipy extension module, telling disutils which Fortran DLLs they depend on?

Correct. It's not just the DLLs that we make, but the DLLs in the MinGW folder too.

matthew-brett commented:
Hmm.  For the DLLs in the mingw folder, how about using mingwpy or one of the newer static mingw variance that @carlkl has been building?  I'm worried about putting the mingw DLLs on the path, with their current names, I guess we could easily run into a DLL clash with other code linking against these DLLs.

Similarly, I wonder if there's a way to keeping the new Fortran DLL names unique, maybe with a hash of their object file contents?  Just in case something else is using these DLL names?

xoviat commented:
There are no issues with the fortran DLLs that we make clashing. In addition, the PATH is only modified for scipy, so there should not be clashes if scipy doesn't spawn subprocesses.

xoviat commented:
Also I don't think gfortran is called with -O3. We should at least try to run as fast as possible.

matthew-brett commented:
I am happy to be corrected, but I believe that we'll run into trouble if any Python module imports a DLL with the same name, either before Scipy imports (Scipy will get the wrong DLL) or after it imports (the other module will get the wrong DLL).  It might not be happening now, but it's difficult to predict if it will happen in the future.

xoviat commented:
Yes, if it the same DLL. However, I do believe that the MinGW DLLs have versioned names.

rgommers commented:
> if scipy doesn't spawn subprocesses.

SciPy itself spawns only subprocesses in the test suite AFAIK.

rgommers commented:
> Approach here is to list all the scipy Fortran libraries inside numpy distutils. It would be better to specify them inside Scipy, perhaps (@pv) in setup.py.

Agreed that it should be in ``scipy``. Preferably outside setup.py I would think - perhaps in a simple plain text file in a format that ``numpy.distutils`` can read reliably when it needs to?

pv commented:
As I understand, it should not be necessary to have the big global 
ordered list. Rather, the order could be specified for each lib separately.

matthew-brett commented:
DLLs again - just to be clear, my understanding is that DLL names are global to the process.  If package P imports DLL `foo`, then, if package Q imports DLL foo, it will get the `foo` that P imported.  Hence the problem about DLL names - it's very easy to generate a clash, and for the clash to cause weird and hard to debug problems.

carlkl commented:
import

all DLLs can be imported into a process only once, selected by the filename of the DLL. Windows DLL search order is: first Windows system libraries, then executable (https://github.com/numpy/numpy/wiki/windows-dll-notes).

Please forget to try the change the DLL search path with the PATH environment variable. This won't work in all cases. I can't recommend the approach of https://github.com/scipy/scipy/pull/7613 at all.

matthew-brett commented:
@carlkl - just to clarify again (I know you meant this) but, for Windows, once you've imported `c:\a_path\foo.dll` then any attempt to import something with basename `foo.dll` will get the contents of `c:\a_path\foo.dll` - for example, loading `c:\another_path\foo.dll` will in fact get `c:\a_path\foo.dll`.

matthew-brett commented:
@carlkl - can you say more about when adding the library directory to the PATH will not work?  And why you can't recommend it?  Did you have a chance to read https://github.com/pypa/wheel-builders/pull/2/files#diff-37a1ef08dd1b0096d769190274121258R378 ?

njsmith commented:
If DLL name collisions are a concern, then [Mach-O Mach-O Mangler](https://github.com/njsmith/machomachomangler) can help you give them unique names. Basically it's the same trick that `auditwheel` uses `patchelf` for, but on Windows: give the DLLs a unique name, and then patch the binaries that refer to them to look for the name name. [Example here](https://github.com/njsmith/machomachomangler#pe-features)

I'd generally recommend giving all DLLs unique names, and then using either the `LoadLibrary` trick or `PATH` manipulation to load them. The advantage of `LoadLibrary` is that it's very local. The advantage of manipulating `PATH` is that you can do it once and then the libraries are loaded on demand as needed, and you can even run native .exe's that are linked against your uniquified names (that's why it shows up in that pynativelib thing) -- but there's also more potential for making mistakes and accidentally affecting the load of libraries you didn't mean to. If you give your DLLs unique names and stick them in an otherwise empty directory then either is pretty safe AFAIK.

matthew-brett commented:
Nathaniel - what is the LoadLibrary trick?

carlkl commented:
Adding the folder to PATH may not work:

- The PATH is more or less the lowest place in the Windows DLL search order. 
- there are corner cases like frozen executables, virtual env, multiprocessing and so on you have to consider.

matthew-brett commented:
@carlkl - could you be more specific about the problems you see?  I guess, if there were somehow two different scipies being imported in the same process this could be a problem - is that what you mean?  In that case - what approach are you suggesting?

njsmith commented:
@matthew-brett: if you call `LoadLibrary("c:\whatever\absolute\path\to\foo.dll")`, then that file gets loaded as `foo.dll` (assuming there isn't already a `foo.dll` loaded), and now if you try to load a library that says it wants `foo.dll`, it uses that one. Even if `c:\whatever\absolute\path\to` was never on any search path.

@carlkl: the search order doesn't matter if your dll is named like `gfortran-lib-3409a0bef0139.dll`, where the gibberish is a hash of the file contents; as long as you can find it at all, then it will definitely be the one you were looking for.

carlkl commented:
@matthew-brett, @njsmith 

using LoadLibrary with an absolute PATH means you have to ensure to get the PATH right. I.e.: you also have to consider frozen executables as well.

matthew-brett commented:
@carllk - sorry to be dumb, but could you unpack for me, what approach you are suggesting, instead of setting the PATH (or loading the library by full path with LoadLibrary)?

njsmith commented:
@carlkl: frozen executables are a pretty unusual case, and handling them likely requires some sort of ad hoc integration with whatever hacks they're using. I don't think we should worry about them until after the basic functionality is working.

carlkl commented:
@matthew-brett @njsmith ,

please test the scipy wheels (https://github.com/scipy/scipy/issues/7597#issuecomment-315886182). No PATH manipulation is needed, as all scipy `pyd` bimaries are distributed in the right folders. You also need to install numpy from here: https://ci.appveyor.com/project/matthew-brett/windows-wheel-builder/build/1.0.57

pv commented:
@carlkl: I guess testing the wheels is less interesting (I can believe 
you if you say they work) than how they were produced. Did you have the 
recipe producing them somewhere?

carlkl commented:
1. I patched `mingw32ccompiler.py` and `gnu.py`  from my local numpy installation. Essentially I added compilerflags for mingw.
2. For cp35 and cp36 I copied `npymath.lib` from the `numpy-cp34` build as `libnpymath.a` into `numpy/core/lib`.

The reason for `2.` is, that `npymath.lib` created by VS2015 is not readable by `binutils`. I believe this is due to the fact, that this library is a mixture of static and dynamic import library. Using a pure static import library with the name `libnpymath.a` ensures, that mingwpy picks up a readable library and VS picks up its own `npymath.lib` library.

Scipy is unpatched and build by (still unpublished) gcc-5.3 version.

matthew-brett commented:
@carlkl - would you mind dropping this into an Appveyor build - maybe patterned after https://github.com/matthew-brett/build-scipy?

xoviat commented:
closing for now to avoid ci runs

matthew-brett commented:
@xoviat - you can also use `[skip ci]` in the commit message to avoid CI runs on both travis and appveyor.

xoviat commented on numpy/distutils/fcompiler/gnu.py:
```diff
@@ -244,6 +250,26 @@ class Gnu95FCompiler(GnuFCompiler):
+    def __init__(self, *args, **kw):
+        GnuFCompiler.__init__(self, *args, **kw)
+        try:
+            with open("runtime_shared_libs.txt") as fp:
```
The only required global lib is actually the OpenBLAS dll, so this should be moved to system_info in the future. However, this is good enough for now.

xoviat commented:
@rgommers Can you review this?

xoviat commented:
As of now, there are no more potential name conflicts because everything is generated with randomized names from the static libs from system_info, including openblas, and the fortran sources.

xoviat committed:
distutils: gnu: patch fcompile

This allows mingw's gfortran to work with MSVC. DLLs are
autogenerated using heuristics that should work with most
cases. In addition, a libopenblas DLL is compiled from
the static lib for use with MSVC.

All generated DLLs have randomized names so that no clashes
will occur.

carlkl commented:
@matthew-brett, concerning possible problems loading DLLs with the help of PATH manipulation see here: https://github.com/numpy/numpy/issues/9272#issuecomment-311191293 and here: https://github.com/numpy/numpy/issues/6923#issuecomment-227241753. 

The approach used by the numpy MKL (now in _distributor_init.py) is to append the numpy/core folder to the PATH environment.  All MKL DLLs loaded directly by _multiarray.pyd will be located alongside _multiarray.pyd in numpy/core. However, subseqent loaded MKL DLLs will be found in the windows folder first, if existent, regardless if numpy/core is defined in PATH or not.

This problem does not directly apply to numpy-openblas, as long as the openblas DLL is linked statically. 

My feeling is, that loading DLLs from a folder defined by the PATH environment may work for most users, but may create hard to find bugs for some corner cases.

xoviat commented:
The other way that it could be done is to load all of the DLLs as soon as scipy is imported. 

> My feeling is, that loading DLLs from a folder defined by the PATH environment may work for most users, but may create hard to find bugs for some corner cases.

Unless we have lazy-load DLLs (which we don't as of now), either the DLL is loaded or it isn't. If it isn't, then you get a pretty loud "DLL load failed" message, which is not at all subtle. And with the randomized names, the chance that you load a DLL that you don't want is 1/36^5.

> This problem does not directly apply to numpy-openblas, as long as the openblas DLL is linked statically.

It's actually your DLL that we're using, so if you say that it is static, then I believe you.

pv commented:
Note that instead of randomized names, you could use the sha1 or other
hash of the content. Then collisions won't occur even if there's some
problem with the seeding of the random generator.

xoviat commented:
We don't know the content of the DLL until we generate the name. 

object files + DLL name =gfortran=> DLL + def file
def file =link.exe=> msvc lib

If you want, we can hash the object files.

xoviat commented:
py34 still runs into #4101 apparently.

njsmith commented:
You could hash the dll file, then rename it, and modify the .def file before generating the .lib. The .def file is just plain text and has the dll name listed at the top, it's easy to edit.

xoviat commented:
The def file is easy to edit but I don't have code that can currently parse it and the format could change in the future. The main problem right now is the MANIFEST failure on msvc 2010.

xoviat commented:
The should be OK to merge:
- Does not break any existing functionality apparently or theoretically (as modified functionality was not already supported)
- Allows new functionality on any python 2.7; 3.4-3.6 assuming that msvc can compile C code correctly.

pv commented:
Without whitespace changes: https://github.com/numpy/numpy/compare/master...pv:pr-9431

pv commented on numpy/distutils/misc_util.py:
```diff
@@ -2287,6 +2299,16 @@ def generate_config_py(target):
+
+import os
+
+os.environ["PATH"] += os.pathsep + os.path.join(os.path.dirname(__file__), '_lib')
```
Probably should't add this --- generally there's no `_lib` directory.

xoviat replied:
This is only added when DLLs are generated and the generation code places the DLLs in the _lib directory and creates the lib directory if it does not exist.

All of this only happens if there is fortran code and mingw is available and ifortran is not available and the compiler is msvc. In other words, if there is not other option but to create the lib directory and place the DLLs in it.

xoviat commented on numpy/distutils/command/build_ext.py:
```diff
@@ -256,6 +260,18 @@ def run(self):
+        shared_libs = system_info.shared_libs
+        if shared_libs:
+            runtime_lib_dir = os.path.join(
+                self.build_lib, self.distribution.get_name(), '_lib')
```
@pv Here you go.

pv replied:
I see -- I grepped the sources but didn't notice that.

pv commented:
A very good thing about this approach is that (i) it is not so difficult to set up a working toolchain, you can just download msvc from microsoft and gfortran e.g. from anaconda, and (ii) the whole toolchain is maintained by someone else.

I've been fiddling with this now a little bit.  https://github.com/numpy/numpy/compare/master...pv:pr-9431-2

Some observations:

- Linking all object DLLs to all pyd extensions seems a bit ugly --- the link order of the DLL libraries appears to be unspecified, so it is not that clear you are guaranteed to get right symbols. I'm not sure if this is the cause for problems on Python 3.6 in the scipy appveyor builds, but if it is, then I guess something more selective has to be done.
- Using `system_info.shared_libs` assumes the build runs in one go, which isn't always true. This information should be recorded on disk, so that builds can be resumed and be idempotent.
- The openblas msvc magic relies on a side effect of link_wrapper_lib, which adds the openblas lib to the chained DLLs.

It would be possible to do this a bit differently (cf the branch linked above, which at least locally on python.org/py3.6+VS2017/msvc+mingw-64/gfortran manages to build scipy that passes the test suite, numpy test suite has 3 failures on complex trig function branch cuts, so looks good):

- Give up building static libraries from fortran sources if the linker doesn't understand Fortran compiler object files. If there's no way to have DLLs with undefined symbols (didn't try what `-Wl,--allow-shlib-undefined` does...), it seems messy to try to work around this with wrapper DLLs because, as you noted, build_clib just does not have the necessary information.
- Since `numpy.distutils` controls both `build_clib` and `build_ext`, push the concept of incompatible object files to this level.
- Use "fake" static libraries instead of real static libs when needed, just lists of object files.
- Openblas treatment is now simplified to just considering "openblas.a" just as another unlikable static lib.
- Record the status of "extra-dlls" on disk, rather than in system_info, so that builds can be resumed.

Thoughts?

carlkl commented:
scipy *can* be compiled with mingwpy without the use of MSVC, see https://github.com/scipy/scipy/pull/7616#issuecomment-317975247.  Some minor patches for numpy are mandantory (nor PR right now). Mingwpy needs to be customized to allow linkage against the correct CRT runtime. An unsolved problem right now is the fact that the import library `npymath.lib` can't be read by the GNU linker in case of VC2015. A possible solution would be a second build of this library targetting a different name `libnpymath.a` compiled with the `/MT` instead of the `/MD` flag. As a workaround the library compiled with VC2010 can do the job. With that the MSVC based numpy works for package builders using both compiler toolchains: MSVC *AND* mingw-w64 

The advantage of approach mentioned above is that the original file layout of scipy the wheel is left intact. No randomized pyd-files in a dedicated folder are needed.

The disadvantage right now is that UCRT (cp-3.5 and above) CRT linkage needs some more testing. 

No 64bit errors and failures, thats the good news, The 32bit erorrs and failures are also related to cholesky/linalg. I tested different versions of OpenBLAS with identical results. As I have no time for debugging due to travelling: here are some links for interested testers:
[scipy-0.19.1-cp27-cp27m-win32.whl for use with numpy+openblas](http://040d8cbd078024f25786-9ba0e05766e4bd68e53ad28ecdbfcaa7.r23.cf2.rackcdn.com/scipy-0.19.1-cp27-cp27m-win32.whl)
[scipy-0.19.1-cp27-cp27m-win_amd64.whl for use with numpy+openblas](http://040d8cbd078024f25786-9ba0e05766e4bd68e53ad28ecdbfcaa7.r23.cf2.rackcdn.com/scipy-0.19.1-cp27-cp27m-win_amd64.whl)
[scipy-0.19.1-cp35-cp35m-win32.whl for use with numpy+openblas](http://040d8cbd078024f25786-9ba0e05766e4bd68e53ad28ecdbfcaa7.r23.cf2.rackcdn.com/scipy-0.19.1-cp35-cp35m-win32.whl)
[scipy-0.19.1-cp35-cp35m-win_amd64.whl for use with numpy+openblas](http://040d8cbd078024f25786-9ba0e05766e4bd68e53ad28ecdbfcaa7.r23.cf2.rackcdn.com/scipy-0.19.1-cp35-cp35m-win_amd64.whl)

The numpy wheels are available here: [numpy+openblas](http://040d8cbd078024f25786-9ba0e05766e4bd68e53ad28ecdbfcaa7.r23.cf2.rackcdn.com)

Concerning the MSVC based building approach of @xoviat following comes into my mind:

one should try to put together a static import library out of available static import libraries provided by mingw-w64 containing *exclusively* all missing symbols needed for link gfortran based object files with the MSVC linker. Double symbols should be avoided. A dependency to the gcc runtime DLLs should be avoided as well by using only objects from static import libraries. This may allow to get away from the approach of randomized DLLs for this kind of build as well.

xoviat commented:
> The openblas msvc magic relies on a side effect of link_wrapper_lib, which adds the openblas lib to the chained DLLs. It'll also bomb if you somehow manage to get a precompiled msvc-compatible openblas.

That's not correct is something that I already considered. Wrapper libs are only generated if the openblas lib ends with .a, which is typically only used for mingw. The worst-case scenario is that the same wrapper lib could be re-generated twice if setup.py build was re-called, which would increase the size of the build output.

As a side note, the OpenBLAS project specifically says that they will not release an msvc-compatible lib.

> Linking all object DLLs to all pyd extensions seems a bit ugly --- the link order of the DLL libraries appears to be unspecified, so it is not that clear you are guaranteed to get right symbols. I'm not sure if this is the cause for problems on Python 3.6 in the scipy appveyor builds, but if it is, then I guess something more selective has to be done.

You are correct. This is a terrible approach, [but I already noted that](https://github.com/scipy/scipy/pull/7616#issuecomment-320450038):

> You're somewhat correct here. The problem is the dependencies are not known until build_ext but they need to be known in build_clib in order to use the correct approach. build_ext already has initialize_command (or whatever its called) build_clib. I'm not sure whether introducing a circular initialize_command (with build_clib initializing build_ext as well) would be okay, but the "traditional" distutils model becomes a nightmare pretty quickly here.

pv commented:
Do you then have comments on how to move this forward? If the DLL link order matters and is the cause for the failures, the approach currently in this PR is incorrect and something else should be done.

Whether scipy can be built with mingwpy is probably somewhat orthogonal to this PR, because it only concerns the msvc+gfortran combination-

xoviat commented:
> one should try to put together a static import library out of available static import libraries provided by mingw-w64 containing exclusively all missing symbols needed for link gfortran based object files with the MSVC linker. Double symbols should be avoided. A dependency to the gcc runtime DLLs should be avoided as well by using only objects from static import libraries. This may allow to get away from the approach of randomized DLLs for this kind of build as well.

MinGW DLLs should not be statically linked with MSVC. This is not officially supported and can quickly become a maintenance nightmare.

carlkl commented:
I think it is worth to invest some effort to create and use an additional mingw-w64 based dedicated static import library for this purpose as I described in https://github.com/numpy/numpy/pull/9431#issuecomment-320531115.
If the MSVC link command is able to link all this together it should be possible to create all `pyd`-files without further dependencies on non MSVC libraries (beside OpenBLAS) and place it on its original place.

xoviat commented:
>  If the DLL link order matters and is the cause for the failures, the approach currently in this PR is incorrect and something else should be done.

This PR makes the link order deterministic, or it should.

xoviat commented:
> I think it is worth to invest some effort to create and use an additional mingw-w64 based dedicated static import library for this purpose as I described in #9431 (comment).

Get appveyor working then!

pv commented:
The wrapper DLL files are in unspecified order in the gfortran call, since it's `list(set(...))` so easily fixable. If ordered, is the build order for static libraries is the right order for linking the extensions files?

Also, other things being equal, I would prefer to not link everything to everything, which is also feasible as demonstrated.

xoviat commented:
> The wrapper DLL files are in unspecified order in the gfortran call, since it's list(set(...)) so easily fixable. If ordered, is the build order for static libraries is the right order for linking the extensions files?

Good catch!

> Also, other things being equal, I would prefer to not link everything to everything, which is also feasible as demonstrated.

Definitely feasible, but this information is contained in the extensions list and needs to be present in build_clib. How to do that?

xoviat commented:
> Definitely feasible, but this information is contained in the extensions list and needs to be present in build_clib. How to do that?

More specifically, how to do that in a 'sane' manner?

xoviat commented:
@pv Sorry for not looking at your PR first. I will look at that and then get back to this.

pv commented:
Yes, build_clib is certainly a problem and I'm not sure if there's a beautiful solution to this.
I also came across `-Wl,--allow-shlib-undefined`, but didnt' try to play with it --- if this it actually works, it perhaps could postpone defining the link order to build_ext in a better way.

xoviat commented:
@pv So merging your changes, I see this:

```python
<<<<<<< HEAD
        if c.compiler_type == "msvc" and not self.create_msvc_openblas_lib(info):
            return False

=======
>>>>>>> pv/pr-9431-2
```

As far as I can tell this may conflict with your approach to defer linking objects. The problem is that we need to present a unified external interface with system_info, which means using the MinGW compatible static lib which cannot be linked to MSVC without a wrapper lib. What approach would you suggest here?

pv committed:
distutils: handle unlinkable object files in build_clib/build_ext, not gnu

Add concept of unlinkable Fortran object files on the level of
build_clib/build_ext.

Make build_clib generate fake static libs when unlinkable object files are
present, postponing the actual linkage to build_ext.

This enables MSVC+gfortran DLL chaining to only involve those DLLs that
are actually necessary for each .pyd file, rather than linking everything
in to every file. Linking everything to everywhere has issues due to
potential symbol clashes and the fact that library build order is
unspecified.

Record shared_libs on disk instead of in system_info. This is necessary
for partial builds -- it is not guaranteed the compiler is actually called
for all of the DLL files.

Remove magic from openblas msvc detection. That this worked previously
relied on the side effect that the generated openblas DLL would be added
to shared_libs, and then being linked to all generated outputs.

xoviat committed:
distutils: system_info: fixup

xoviat commented:
Scipy builds appear to work well (except for 32) with the new changes. Is there anything else holding up this PR?

pv commented on numpy/distutils/fcompiler/gnu.py:
```diff
@@ -335,32 +342,148 @@ def get_libraries(self):
+        if c_compiler.compiler_type != "msvc":
+            raise ValueError("This method only supports MSVC")
+
+        object_hash = self._hash_files(objects)
```
I see now this should actually be `self._hash_files(list(objects) + list(chained_dlls))`.

pv reviewed (approved):
This looks ok to me, and I don't currently have suggestions for improvements, modulo the one comment above.

It's still of course a bit of a workaround, but adding one more working method for compilation on windows should still be useful.

pv commented:
NB. the rebase of the branch looks a bit strange now, with duplicate commits.

xoviat commented:
I will squash the commits.

xoviat committed:
distutils: gnu: hash files + chained dlls

xoviat commented:
/cc @rgommers

rgommers commented on numpy/distutils/fcompiler/gnu.py:
```diff
@@ -193,13 +199,10 @@ def get_libraries(self):
             # the following code is not needed (read: breaks) when using MinGW
             # in case want to link F77 compiled code with MSVC
             opt.append('gcc')
-            runtime_lib = msvc_runtime_library()
```
Question: this change is in the class for ``g77``, so is this a cleanup that looks right or have you been using ``g77``?

Comments:
- the code comment above about MinGW looks like it applied to these lines and should be removed.
- the import of ``msvc_runtime_library`` is now unused and should be removed.

xoviat replied:
These lines actually had a signfiicant impact becuase the class is inherited. It's not just a "cleanup"

rgommers replied:
Ah right, clear.

rgommers commented on numpy/distutils/fcompiler/gnu.py:
```diff
@@ -256,25 +259,29 @@ def version_match(self, version_string):
+        'linker_so': ["<F90>", "-Wall", "-g"],
+        'archiver': ["ar", "-cr"],
+        'ranlib': ["ranlib"],
+        'linker_exe': [None, "-Wall"]
```
this formatting change is for the worse, the old version was easier to read - can you undo?

xoviat replied:
I will revert this section.

xoviat committed:
distutils: gnu: revert :art:

rgommers commented:
> /cc @rgommers

I had a quick read through. Did not see anything critical that should prevent merging or break other configs. Traveling at the moment, so haven't been able to follow all the discussions on Windows builds in detail.

xoviat committed:
distutils: gnu: cleanup

xoviat committed:
distutils: gnu: revert :art:

xoviat commented:
FYI. 

> Limitations: fortran code that needs symbols from C code provided by the
project cannot be compiled like --- this would require using two C compilers
(gcc + msvc) and this is getting too complicated, so it's not supported
(will fail with undefined symbol errors at link time). However, the typical
case of wrapping Fortran codes to Python doesn't need this.

This is actually not correct. Supporting this (which I'm not going to get into this point) would require one additional call to lib.exe in build_ext. It is something that could be implemented with moderate additional effort. See [here](https://docs.microsoft.com/en-us/cpp/build/reference/using-an-import-library-and-export-file) for more details

charris commented:
Needs a release note in `doc/release/1.14.0-notes.rst`.

pv committed:
DOC: release: add mention of msvc+gfortran to release notes

pv commented:
Release notes added. There's a merge conflict, so I'll rebase and push this.

xoviat commented:
Thanks so much for taking care of the merge conflicts.

charris commented:
Lets get this in. If nothing else it is a good starting point for further fixes if needed. Thanks @xoviat and the reviewers.
